### ISSUE TITLE
don't warn about group/world-readable KUBECONFIG if it contains no secrets

### ISSUE THREAD

**Issue by flokli** [helm/helm#11083]
helm currently warns on every command:

```
WARNING: Kubernetes configuration file is group-readable. This is insecure. Location: …
WARNING: Kubernetes configuration file is world-readable. This is insecure. Location: …
```

The warning isn't very useful.

In my case, the kubeconfig uses OIDC for logins, so it doesn't contain any secrets.
Instead of embedding client key data certs (via `client-key-data`), `kubeconfig` might also contain paths to files containing secrets (via `client-key`). Permissions for these are not checked either.

This check should probably simply be removed. It produces both false positives and false negatives, and understanding `kubeconfig` shouldn't be a concern of `helm` either, but of the Kubernetes SDK.

related: https://github.com/helm/helm/issues/9115

**Comment by yxxhero** [comment]
@flokli This is only a reminder. you can change the permission of KUBECONFIG to remove it.

**Comment by jetersen** [comment]
@yxxhero Yes you can change the permissions of kubeconfig if you follow the extra steps: https://github.com/helm/helm/issues/9115#issuecomment-1137431441

Might be easy on some OSes but Windows and WSL requires extra step also when tools like rancher-desktop suggest to symlink the kubeconfig between Windows and WSL so you do not have to maintain two configs.

By default in WSL it cannot see file permission metadata on Windows files you have to enable that.

**Comment by sudermanjr** [comment]
Personally I think that an environment variable to suppress these warnings is a cleaner approach than simply removing them. Something along the lines of `SUPPRESS_KUBECONFIG_PERMS_WARNING` or something explicit like that.

**Comment by flokli** [comment]
The main case here is that the `kubeconfig` files doesn't always contain secrets:

 - Secrets can either live in other files, that are referred to from `kubeconfig` (`client-key` instead of `client-key-data`)
 - Secrets can not exist at all (OIDC)

It shouldn't be a concern of helm to try to understand all the various formats.

**Comment by jetersen** [comment]
if it is not a concern for helm than it should remove the readable check.
Or at the very least provide a way to suppress the warning.

**Comment by colin-arnott-xero** [comment]
as the description implies, this check is quite flaky, thus if you have the flexibility to use the `KUBECONFIG` environment variable, providing the "insecure" one second[0] will work around this check without making any permission changes.

0] you can even use an empty first file `KUBECONFIG=":path/to/insecure"`

**Comment by flokli** [comment]
Yeah, as written above, this should be a Kubernetes SDK concern, so all libraries connecting to kubernetes could warn in these cases. But having it downstream in `helm` doesn't make too much sense, especially considering it's often wrong.

**Comment by flokli** [comment]
I'm aware it's a warning/reminder that can be ignored. The "this is insecure" is a bit of a too strong message, and in many cases wrong - which is why I opened this issue. Of course setting the permissions might be a workaround - at least in filesystems supporting this, but I'd rather prefer if helm wouldn't try to make guesses that are wrong quite often...

**Comment by joejulian** [comment]
Looking at the history of that function, it's been stated rather explicitly by the core maintainers that this is a result of a CNCF security team audit.

> To be clear, will not accept PRs to suppress the security warning. This issue was pointed out by a security audit sponsored by the CNCF, and the warning was the security team’s recommendation.
> 
> However, if there are ways where we can identify that a) the kubeconfig is mounted in an NTFS drive, and b) permissions are indeed restricted to the user, we would welcome PRs that would address this case to remove the warning.

https://github.com/helm/helm/issues/8776#issuecomment-812146980


Based on that second paragraph, I suspect logic that would also determine that there were no secrets in the kubeconfig would also be considered.

**Comment by colin-arnott-xero** [comment]
not stale

**Comment by chrisdoherty4** [comment]
> Looking at the history of that function, it's been stated rather explicitly by the core maintainers that this is a result of a CNCF security team audit.
> 
> > To be clear, will not accept PRs to suppress the security warning. This issue was pointed out by a security audit sponsored by the CNCF, and the warning was the security team’s recommendation.
> > However, if there are ways where we can identify that a) the kubeconfig is mounted in an NTFS drive, and b) permissions are indeed restricted to the user, we would welcome PRs that would address this case to remove the warning.
> 
> [#8776 (comment)](https://github.com/helm/helm/issues/8776#issuecomment-812146980)
> 
> Based on that second paragraph, I suspect logic that would also determine that there were no secrets in the kubeconfig would also be considered.

That seems a little inconsistent: `kubectl` and other core Kubernetes tooling doesn't warn, why is Helm?

**Comment by joejulian** [comment]
I don't think we can speculate about why something didn't happen.

**Comment by chrisdoherty4** [comment]
> I don't think we can speculate about why something didn't happen.

I think we can and should because it'll lead to all the right answers [sarcasm] :).

---

On a more relevant note, do you think an env var, as others have suggested, to suppress would be humorable?

**Comment by joejulian** [comment]
I kind-of doubt they'd go for that. Anything that could be done out-of-sight of the user that could make it possible to surreptitiously install software onto your cluster is probably going to be rejected. I think the most likely thing they'd accept, if at all, is a flag.

**Comment by NoseyNick** [comment]
I'm upvoting this. Like (presumably) every other EKS user, my kubeconfig contains `exec: command: aws args: --region BLAH eks get-token --cluster-name BLAH` and no secrets whatsoever. I also have really good group-collaboration reasons why my group wants/**needs** to share their kubeconfigs **with group permissions**. This is not an "NTFS" thing, this is a deliberate, well-considered decision. I'd even go so far as to say that's what group (or indeed world) permissions are **FOR**, isn't it?

> To be clear, will not accept PRs to suppress the security warning. This issue was pointed out by a security audit sponsored by the CNCF, and the warning was the security team’s recommendation.

Can we at least point out that the CNCF have been misled into thinking that this file is necessarily sensitive? Why **NOT** support an option in that kubeconfig to say something like "I hereby declare that this file is not secret or sensitive" for example? Give me the option to make it clear that this was a deliberate, well-considered decision.

My `/etc/passwd` file is also group/world readable. My `/etc/shadow` file is group (shadow) readable! I have **group** mailboxes which are deliberately **group** readable, and contain far more sensitive info than any of my kubeconfigs. C'mon, have some sense of proportion here.

**Comment by bigwheels16** [comment]
I'm not sure why this is a problem.  I don't see what the big deal is inundating users with spurious warning messages, to the point that they can't tell the difference between real warnings and fake ones.

...I know some of you might say that we're actually teaching users to just ignore warning messages altogether, and to you "security" professionals I would say that that is a purely academic discussion and that's where it should stay.  This would never happen to users in the real world.  They are way too smart for that.

**Comment by NikPaushkin** [comment]
@bigwheels16 
> I'm not sure why this is a problem. I don't see what the big deal is inundating users with spurious warning messages

The problem is that the one who executes `helm` might be not a human, but another program. So this useless warning makes the subprocess running tough and ugly.

**Comment by flokli** [comment]
Not stale, people again got bitten by this, see https://github.com/helm/helm/pull/12316#issuecomment-1732527352

**Comment by flokli** [comment]
Still relevant.

**Comment by silopolis** [comment]
Heavily pondering on @sudermanjr proposal of a quietening environment variable.

I need my KUBECONFIG to be group-readable, so my vagrant user can be in the minikube group with rw access to it.

**Comment by flokli** [comment]
> I saw this error on `v3.12.1`. After upgrading to `v3.14.1`, the error disappeared

How should it? The check is still there: https://github.com/helm/helm/blob/51a07e7e78cba05126a84c0d890851d7490d2e20/cmd/helm/root_unix.go#L53

Still relevant, also see https://github.com/helm/helm/pull/12316#issuecomment-1732527352

**Comment by flokli** [comment]
Still relevant.
